runelite-client/src/main/java/net/runelite/client/plugins/puzzlesolver/solver/monkeymadness/PuzzleStateMM.java

```diff
@@ -0,0 +1,855 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.puzzlesolver.solver.monkeymadness;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import net.runelite.api.Point;
+import net.runelite.client.plugins.puzzlesolver.solver.PuzzleSolver;
+import static net.runelite.client.plugins.puzzlesolver.solver.PuzzleSolver.BLANK_TILE_VALUE;
+import static net.runelite.client.plugins.puzzlesolver.solver.PuzzleSolver.DIMENSION;
+import net.runelite.client.plugins.puzzlesolver.solver.PuzzleState;
+import net.runelite.client.plugins.puzzlesolver.solver.pathfinding.Pathfinder;
+
+public class PuzzleStateMM extends PuzzleState
+{
+	private int[] pieces;
+	private int emptyPiece = -1;
+
+	private List<List<Integer>> validRowNumbers = new ArrayList<>();
+	private List<List<Integer>> validColumnNumbers = new ArrayList<>();
+
+	private List<Point> swaps = new ArrayList<>();
+
+	public PuzzleStateMM(int[] pieces)
+	{
+		super(pieces);
+
+		this.pieces = super.getPuzzle();
+		this.emptyPiece = super.getEmptyPiece();
+
+		//Add valid numbers for rows and columns
+		validRowNumbers.add(Arrays.asList(0, 1, 2, 3, 4));
+		validRowNumbers.add(Arrays.asList(6, 7, 8, 9));
+		validRowNumbers.add(Arrays.asList(12, 13, 14));
+		validRowNumbers.add(Arrays.asList(18, 19));
+		validColumnNumbers.add(Arrays.asList(5, 10, 15, 20));
+		validColumnNumbers.add(Arrays.asList(11, 16, 21));
+		validColumnNumbers.add(Arrays.asList(17, 22));
+		validColumnNumbers.add(Arrays.asList(23));
+	}
+
+	PuzzleState solvePathfinder(PuzzleState preSolved, Pathfinder pathfinder)
+	{
+		PuzzleSolver solver = new PuzzleSolver(pathfinder, new PuzzleState(preSolved));
+		solver.run();
+
+		if (solver.hasFailed())
+		{
+			return null;
+		}
+
+		return solver.getStep(solver.getStepCount() - 1);
+	}
+
+	void solveRow(int row)
```

Comment:
Don't solve the puzzle in the puzzle state classes. Solving should be in the puzzle solver.